pkg/storage/jetstream/jetstream.go

```diff
@@ -0,0 +1,134 @@
+package jetstream
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/iancoleman/strcase"
+	"github.com/nats-io/nats.go"
+	corev1 "github.com/rancher/opni/pkg/apis/core/v1"
+	"github.com/rancher/opni/pkg/config/v1beta1"
+	"github.com/rancher/opni/pkg/logger"
+	"github.com/rancher/opni/pkg/storage"
+	"go.uber.org/zap"
+)
+
+const (
+	tokensBucket       = "tokens"
+	clustersBucket     = "clusters"
+	keyringsBucket     = "keyrings"
+	rolesBucket        = "roles"
+	roleBindingsBucket = "rolebindings"
+	dynamicBucket      = "dynamic"
+)
+
+type JetStreamStore struct {
+	JetStreamStoreOptions
+
+	nc *nats.Conn
+	js nats.JetStreamContext
+
+	kv struct {
+		Tokens       nats.KeyValue
+		Clusters     nats.KeyValue
+		Keyrings     nats.KeyValue
+		Roles        nats.KeyValue
+		RoleBindings nats.KeyValue
+	}
+	logger *zap.SugaredLogger
+}
+
+var _ storage.Backend = (*JetStreamStore)(nil)
+
+type JetStreamStoreOptions struct {
+	BucketPrefix string
+}
+
+type JetStreamStoreOption func(*JetStreamStoreOptions)
+
+func (o *JetStreamStoreOptions) apply(opts ...JetStreamStoreOption) {
+	for _, op := range opts {
+		op(o)
+	}
+}
+
+func NewJetStreamStore(ctx context.Context, conf *v1beta1.JetStreamStorageSpec, opts ...JetStreamStoreOption) (*JetStreamStore, error) {
+	options := JetStreamStoreOptions{
+		BucketPrefix: "gateway",
+	}
+	options.apply(opts...)
+
+	lg := logger.New(logger.WithLogLevel(zap.WarnLevel)).Named("jetstream")
+
+	nkeyOpt, err := nats.NkeyOptionFromSeed(conf.NkeySeedPath)
+	if err != nil {
+		return nil, err
+	}
+	nc, err := nats.Connect(conf.Endpoint, nkeyOpt, nats.MaxReconnects(-1))
+	if err != nil {
+		return nil, err
+	}
+	js, err := nc.JetStream()
+	if err != nil {
+		return nil, err
+	}
+
+	store := &JetStreamStore{
+		JetStreamStoreOptions: options,
+		nc:                    nc,
+		js:                    js,
+		logger:                lg,
+	}
+
+	store.kv.Tokens = store.upsertBucket(tokensBucket)
+	store.kv.Clusters = store.upsertBucket(clustersBucket)
+	store.kv.Keyrings = store.upsertBucket(keyringsBucket)
+	store.kv.Roles = store.upsertBucket(rolesBucket)
+	store.kv.RoleBindings = store.upsertBucket(roleBindingsBucket)
+
+	return store, nil
+}
+
+func (s *JetStreamStore) upsertBucket(name string) nats.KeyValue {
+	bucketName := fmt.Sprintf("%s-%s", s.BucketPrefix, name)
+	kv, err := s.js.KeyValue(bucketName)
+	if err != nil {
+		if errors.Is(err, nats.ErrBucketNotFound) {
+			kv, err = s.js.CreateKeyValue(&nats.KeyValueConfig{
```

Comment:
`CreateKeyValue` is idempotent.  If the bucket already exists it will throw an error if the config is different to the existing bucket.  It may be simpler to just call create every time.

Reply:
It is not documented to be idempotent, and there is a warning above `nats.KeyValueManager` stating it is experimental and the functionality may change. Without an API guarantee, it is probably safer  to initialize the bucket this way, in case they change the implementation details later.